<?xml version="1.0"?>

<launch>
  <arg name="node_required" default="false"      />
  <arg name="node_name"     default="opensim_ik" />

  <!-- General Parameters -->
  <arg name="n_threads"                default="4"     />
  <arg name="use_marker_positions"     default="false" />
  <arg name="use_link_orientations"    default="false" />
  <arg name="model_path"               default=""      />
  <arg name="model_scaling"            default="false" />
  <arg name="model_calibration"        default="true"  />
  <arg name="input_topic"              default=""      />
  <arg name="out_joint_state_topic"    default=""      />
  <arg name="out_skeleton_group_topic" default=""      />

  <!-- Model Scaler Parameters -->
  <arg name="scale_config_path" default=""   />
  <arg name="scale_initial_t"   default="-1" />
  <arg name="scale_final_t"     default="-1" />

  <!-- IMU Placer Parameters -->
  <arg name="heading_correction"    default="false" />
  <arg name="base_imu_label"        default=""      />
  <arg name="base_heading_axis"     default=""      />
  <arg name="use_marker_based_ik_as_initial_pose" default="false" />
  <arg name="save_calibrated_model" default="false" />
  <arg name="visualize_calibration" default="false" />

  <!-- IK Tool Parameters -->
  <arg name="accuracy"                     default="1e-4"  />
  <arg name="markers_weight"               default="1."    />
  <arg name="orientations_weight"          default="1."    />
  <arg name="sensor_to_opensim_rotation_x" default="0"     />
  <arg name="sensor_to_opensim_rotation_y" default="0"     />
  <arg name="sensor_to_opensim_rotation_z" default="0"     />
  <arg name="visualize_ik"                 default="false" />


  <node pkg="hiros_opensim_ik_solver" type="hiros_opensim_ik_solver" name="$(arg node_name)" output="screen" required="$(arg node_required)">
    <param name="node_name" type="str" value="$(arg node_name)" />

    <!-- General Parameters -->
    <param name="n_threads"                type="int"  value="$(arg n_threads)"                />
    <param name="use_marker_positions"     type="bool" value="$(arg use_marker_positions)"     />
    <param name="use_link_orientations"    type="bool" value="$(arg use_link_orientations)"    />
    <param name="model_path"               type="str"  value="$(arg model_path)"               />
    <param name="model_scaling"            type="bool" value="$(arg model_scaling)"            />
    <param name="model_calibration"        type="bool" value="$(arg model_calibration)"        />
    <param name="input_topic"              type="str"  value="$(arg input_topic)"              />
    <param name="out_joint_state_topic"    type="str"  value="$(arg out_joint_state_topic)"    />
    <param name="out_skeleton_group_topic" type="str"  value="$(arg out_skeleton_group_topic)" />


    <!-- Model Scaler Parameters -->
    <param name="scale_config_path" type="str"    value="$(arg scale_config_path)" />
    <param name="scale_initial_t"   type="double" value="$(arg scale_initial_t)"   />
    <param name="scale_final_t"     type="double" value="$(arg scale_final_t)"     />

    <!-- IMU Placer Parameters -->
    <param name="heading_correction"    type="bool" value="$(arg heading_correction)"    />
    <param name="base_imu_label"        type="str"  value="$(arg base_imu_label)"        />
    <param name="base_heading_axis"     type="str"  value="$(arg base_heading_axis)"     />
    <param name="use_marker_based_ik_as_initial_pose" type="bool" value="$(arg use_marker_based_ik_as_initial_pose)" />
    <param name="save_calibrated_model" type="bool" value="$(arg save_calibrated_model)" />
    <param name="visualize_calibration" type="bool" value="$(arg visualize_calibration)" />

    <!-- IK Tool Parameters -->
    <param name="accuracy"                     type="double" value="$(arg accuracy)"                     />
    <param name="markers_weight"               type="double" value="$(arg markers_weight)"               />
    <param name="orientations_weight"          type="double" value="$(arg orientations_weight)"          />
    <param name="sensor_to_opensim_rotation_x" type="double" value="$(arg sensor_to_opensim_rotation_x)" />
    <param name="sensor_to_opensim_rotation_y" type="double" value="$(arg sensor_to_opensim_rotation_y)" />
    <param name="sensor_to_opensim_rotation_z" type="double" value="$(arg sensor_to_opensim_rotation_z)" />
    <param name="visualize_ik"                 type="bool"   value="$(arg visualize_ik)"                 />
  </node>
</launch>
